engine/physics.3
.\"	$Csoft: physics.3,v 1.10 2003/03/23 11:35:49 vedge Exp $
.\"
.\" Copyright (c) 2002, 2003 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd November 23, 2002
.Dt PHYSICS 3
.Os
.ds vT Agar API Reference
.ds oS Agar 1.0
.Sh NAME
.Nm physics
.Nd agar movement interface
.Sh SYNOPSIS
.Bd -literal
#include <engine/physics.h>
.Ed
.Sh DESCRIPTION
The agar
.Nm
interface manages movement of objects, usually in reaction to input device
events.
The directions are represented by these flags:
.Bd -literal
#define DIR_UP		0x01
#define DIR_DOWN	0x02
#define DIR_LEFT	0x04
#define DIR_RIGHT	0x08
#define DIR_ALL		0xff
.Ed
.Pp
.Sh GENERIC DIRECTIONS
These functions implement generic two-dimensional movement.
.Pp
.nr nS 1
.Ft "void"
.Fn gendir_init "struct gendir *dir"
.Pp
.Ft "void"
.Fn gendir_set "struct gendir *dir" "int direction"
.Pp
.Ft "void"
.Fn gendir_unset "struct gendir *dir" "int direction"
.Pp
.Ft "int"
.Fn gendir_move "struct gendir *dir"
.Pp
.Ft "void"
.Fn gendir_postmove "struct gendir *dir" "int moved"
.nr nS 0
.Pp
The
.Fn gendir_init
function initializes a generic direction.
.Pp
The
.Fn gendir_set
function sets the direction of
.Fa dir
to
.Fa direction .
The
.Fn gendir_unset
requests
.Fa dir
to no longer move in
.Fa direction .
.Pp
The
.Fn gendir_move
function is usually called from a timer callback to update the direction
.Fa dir ,
returning the direction towards which the object should move, or 0 if it
should remain in its current position.
.Pp
If the timer callback has moved object, it should call
.Fn gendir_postmove
with the return value of
.Fn gendir_move
as the
.Fa moved
argument, to effect a
.Fn gendir_unset
operation and increment the necessary counters.
.Sh MAP DIRECTIONS
This interface implements movement of objects between adjacent map nodes.
Game objects are bound to an input device with
.Fn object_control
and the
.Xr input 3
subsystem calls
.Fn mapdir_set
and
.Fn mapdir_unset
in reaction to input device events.
The particular object implementation then calls
.Fn mapdir_move
(usually from a timer callback function) in order to effect movement.
.nr nS 1
.Ft "void"
.Fn mapdir_init "struct mapdir *dir" "struct object *ob" "struct map *map" \
                "int flags" "int speed"
.Pp
.Ft "void"
.Fn mapdir_set "struct mapdir *dir" "int direction"
.Pp
.Ft "void"
.Fn mapdir_unset "struct mapdir *dir" "int direction"
.Pp
.Ft "int"
.Fn mapdir_move "struct mapdir *dir" "int *map_x" "int *map_y"
.Pp
.Ft "void"
.Fn mapdir_postmove "struct mapdir *dir" "int *map_x" "int *map_y" \
                    "int moved"
.nr nS 0
.Pp
The
.Fn mapdir_init
function initializes a map direction for object
.Fa ob
on map
.Fa map .
The
.Fa flags
may contain:
.Pp
.Bd -literal
#define DIR_SCROLLVIEW	0x01	/* Scroll view if we reach boundaries. */
#define DIR_SOFTSCROLL	0x02	/* Animate move from node to node. */
#define DIR_STATIC	0x04	/* Don't change sprites with directions. */
#define DIR_PASSTHROUGH	0x08	/* Pass through map nodes. */
.Ed
.Pp
The
.Fn mapdir_set
function sets the direction of
.Fa dir
to
.Fa direction .
The
.Fn mapdir_unset
requests
.Fa dir
to no longer move in
.Fa direction .
.Pp
The
.Fn mapdir_move
function is usually called from a timer callback to update the direction
.Fa dir ,
returning the direction towards which the object should move, or 0 if it
should remain in its current position.
The
.Fa map_x
and
.Fa map_y
arguments should point to the current coordinates of the object on the map,
and will be updated accordingly.
.Pp
If the timer callback has moved the map reference, it should call
.Fn mapdir_postmove
with the return value of
.Fn mapdir_move
as the
.Fa moved
argument, to effect a
.Fn mapdir_unset
operation, increment the necessary counters or ensure the object keeps
moving.
The
.Fa map_x
and
.Fa map_y
arguments are the same pointers that were passed to
.Fn mapdir_move .
.Sh SEE ALSO
.Xr agar 3 ,
.Xr map 3 ,
.Xr rootmap 3
.Sh HISTORY
The
.Nm
interface first appeared in Agar 1.0


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.0] - Unreleased
### Added
- [**Style Editor**](https://libagar.org/man3/AG_StyleEditor): Display computed style rules in addition to per-instance overrides.
- [**AG_Button**](https://libagar.org/man3/AG_Button): New style attributes `text-align` and `vertical-align`.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Introduce the `style()` method, called by the style compiler to parse widget-specific style attributes into an efficient binary representation.
- [**AG_Object**](https://libagar.org/man3/AG_Object): New function `AG_AccessVariable_NoLock()` (lockless variant of `AG_AccessVariable()`).
- [**AG_Object**](https://libagar.org/man3/AG_Object): Document the `AG_OBJECT_BOUND_EVENTS` flag.

### Removed
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove the "window-detached" event. It is redundant since the generic "detached" event does the same job.

### Changed
- Major performance improvements in the style compiler. Since it runs whenever widgets are resized, this should lead to better fluidity and overall lower CPU usage when Agar windows and container widgets are resized.
- [**AG_Object**](https://libagar.org/man3/AG_Object): In `AG_ObjectDestroy()`, do not call `AG_ObjectDetach()` on child objects. Instead, just cancel any active timers and null their parent pointers without raising a "detached" event. This allows us to perform garbage collection more efficiently.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Make `AG_TLIST_NO_BGLINES` the default.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): In `AG_TlistNewPolledMs()`, accept a `ms` argument of 0 to disable periodical updates (the caller may still trigger updates with "tlist-poll").
- In `AG_WidgetCopyStyle()`, prevent redundant locking by using `AG_AccessVariable_NoLock()`.

### Fixed
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Fix redundant "detached" handler invocations. Improve overall efficiency of the garbage-collection step after windows are destroyed.
- [**AG_Console**](https://libagar.org/man3/AG_Console): Prevent a redundant lock operation in the backend of `AG_ConsoleMsg()`.

## [1.7.1] - 2025-11-15
### Added
- An awesome new Agar logo. Thanks [eightbit](https://8bitprodigy.github.io)!
- Added support for the [Cmake](https://cmake.org) build system.
- Added support for the [Meson](https://mesonbuild.com) build system. Thanks [coleman](https://git.sr.ht/~coleman/)!
- New functions [**AG_InitVideoSDL2**](https://libagar.org/man3/AG_InitVideoSDL2) and [**AG_SetVideoSurfaceSDL2**](https://libagar.org/man3/AG_SetVideoSurfaceSDL2) for integrating with an existing SDL2 display context. Thanks Brigham Keys!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Add "Any File" type.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistCopy()`. Copy all items from a source to a destination `AG_Tlist`.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New constructor `AG_ComboNewFn()`. Sets a `combo-expanded` handler.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New member `nVisItems`. Set the number of items to show by default in expansions.
- [**AG_Checkbox**](https://libagar.org/man3/AG_Checkbox): New functions `AG_CheckboxText()` and `AG_CheckboxTextS()` to update the text label.
- Tweak the default color scheme. Document style defaults as a comment block in `gui/style.css`.
- New Ada bindings for the `Window`, `Box`, `Button`, `Checkbox` and `Combo` widget classes. Updated `ada/gui/demo`.

### Removed
- [**AG_Db**](https://libagar.org/man3/AG_Db): Remove support for DB5.
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove an unnecessary counter variable in `SizeAllocate()`.

### Changed
- Small cosmetic changes to the default color scheme. Set default font size back to 14pt.
- [**AG_Titlebar**](https://libagar.org/man3/AG_Titlebar): Enable the `AG_BUTTON_EXCL` optimization on titlebar buttons.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Allow the dropdown list to be populated by a `combo-expanded` handler.
- [**AG_UCombo**](https://libagar.org/man3/AG_UCombo): Allow the dropdown list to be populated by a `ucombo-expanded` handler.

### Fixed
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Make it again possible to statically initialize `list` before `combo-expanded`. Restores compatibility pre-1.6. Thanks Wally!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix widget geometries not updating when switching to a different Type filter.
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix compilation where `GLOB_TILDE` is unavailable.
- SDL2 drivers: Require at least version 2.0.22 of SDL2 (`SDL_HINT_MOUSE_AUTO_CAPTURE` is necessary for best user experience).
- [**AG_ProgressBar**](https://libagar.org/man3/AG_ProgressBar): Make `padding` work as expected in progress bar. Thanks scaramacai!
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Fix a compilation error.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Fix a keyboard handling error. Don't pass `KeyRelease` events to `XwcLookupString()`. Fixes issue #74. Thanks @tsesani and @scaramacai!
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Introduce the `no-im` (disable input methods) driver option.
- [**MAP**](https://libagar.org/man3/MAP): Fix a crash when attempting to create an object if `pLibsRoot` is NULL.
- [**AG_Web**](https://libagar.org/man3/AG_Web): Fix inline compilation problem due to a typo in `net/web.h`.
- On Darwin platforms, build with `_DARWIN_C_SOURCE` globally. Fixes issue #77. Thanks @kwhr0!
- On FreeBSD check for availability of `CLOCK_SECOND` in addition to `HAVE_CLOCK_GETTIME`.
- Test for the `dirfd()` function at compilation.

engine/physics.c
/*	$Csoft: physics.c,v 1.57 2003/06/29 11:33:41 vedge Exp $	    */

/*
 * Copyright (c) 2001, 2002, 2003 CubeSoft Communications, Inc.
 * <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <engine/engine.h>

#include <engine/map.h>
#include <engine/rootmap.h>
#include <engine/input.h>
#include <engine/physics.h>
#include <engine/view.h>

/*
 * These timings have a granularity proportional to the
 * interval of the timer calling gendir_move().
 */
enum {
	GENDIR_REPEAT_DELAY =	30,	/* Repeat delay */
	GENDIR_REPEAT_IVAL =	10	/* Repeat interval */
};

#ifdef DEBUG
#define DEBUG_MOVE	0x01
#define DEBUG_BLOCKS	0x02

int	physics_debug = DEBUG_MOVE|DEBUG_BLOCKS;
#define engine_debug physics_debug
#endif

void
gendir_init(struct gendir *dir)
{
	dir->set = 0;
	dir->current = 0;
	dir->clear = 0;
	dir->moved = 0;
	dir->offs = 0;
	dir->noffs = 0;
}

void
gendir_set(struct gendir *dir, int direction)
{
	dir->clear &= ~direction;
	dir->set   |=  direction;
}

void
gendir_unset(struct gendir *dir, int direction)
{
	dir->clear |=  direction;
	dir->set   &= ~direction;
}

int
gendir_move(struct gendir *dir)
{
	if (dir->clear != 0) {
		int r = dir->current;

		/* Effect a gendir_unset() operation. */
		dir->current &= ~(dir->clear);
		dir->clear = 0;
		dir->noffs = 0;
		dir->offs = 0;
		return (r);
	}

	if (dir->current != 0) {
		/* Keyboard repeat delay */
		if (dir->noffs > 0) {
			if (dir->noffs++ > GENDIR_REPEAT_DELAY) {
				/* chain */
				return (dir->current);
			}
		}

		/* Repeat interval */
		if (dir->offs++ > GENDIR_REPEAT_IVAL) {
			dir->offs = 0;
			return (dir->current);
		} else {
			return (0);
		}
	} else {
		if (dir->set != 0) {
			dir->current |= dir->set;
			dir->set = 0;
		}
	}
	return (0);
}

void
gendir_postmove(struct gendir *dir, int moved)
{
	if (dir->clear != 0) {
		/* Clear this direction (eg. key release). */
		dir->current &= ~(dir->clear);
		dir->clear = 0;
		dir->noffs = 0;
		dir->offs = 0;
	} else {
		/* Increment the key delay counter. */
		dir->noffs++;
	}
}

void
mapdir_init(struct mapdir *dir, struct object *ob, struct map *map,
    int flags, int speed)
{
	dir->set = 0;
	dir->current = 0;
	dir->clear = 0;
	dir->moved = 0;
	dir->flags = flags;
	dir->speed = speed;
	dir->ob = ob;
	dir->map = map;
}

void
mapdir_set(struct mapdir *dir, int direction)
{
	dir->clear &= ~direction;
	dir->set   |=  direction;
}

void
mapdir_unset(struct mapdir *dir, int direction)
{
	dir->clear |=  direction;
	dir->set   &= ~direction;
}

static void
mapdir_set_motion(struct mapdir *dir, int xmotion, int ymotion)
{
	struct object_position *pos = dir->ob->pos;
	int x, y;

	for (y = pos->y; y < pos->y+pos->submap->maph; y++) {
		for (x = pos->x; x < pos->x+pos->submap->mapw; x++) {
			struct node *node = &pos->map->map[y][x];
			struct noderef *r;

			TAILQ_FOREACH(r, &node->nrefs, nrefs) {
				if (r->layer != pos->z) {
					continue;
				}
				r->r_gfx.xmotion = xmotion;
				r->r_gfx.ymotion = ymotion;
			}
		}
	}
}

static void
mapdir_add_motion(struct mapdir *dir, int xmotion, int ymotion)
{
	struct object_position *pos = dir->ob->pos;
	int x, y;

	for (y = pos->y; y < pos->y+pos->submap->maph; y++) {
		for (x = pos->x; x < pos->x+pos->submap->mapw; x++) {
			struct node *node = &pos->map->map[y][x];
			struct noderef *r;

			TAILQ_FOREACH(r, &node->nrefs, nrefs) {
				if (r->layer != pos->z) {
					continue;
				}
				r->r_gfx.xmotion += xmotion;
				r->r_gfx.ymotion += ymotion;
			}
		}
	}
}

/* Update directions assuming the object is not already moving. */
static void
mapdir_update_idle(struct mapdir *dir)
{
	if (dir->set == 0)
		return;

	if (dir->set & DIR_UP) {
		object_set_submap(dir->ob, "n-move");
		mapdir_set_motion(dir, 0, -1);
		dir->current |= DIR_UP;
		dir->current &= ~(DIR_DOWN|DIR_LEFT|DIR_RIGHT);
	}
	if (dir->set & DIR_DOWN) {
		object_set_submap(dir->ob, "s-move");
		mapdir_set_motion(dir, 0, 1);
		dir->current |= DIR_DOWN;
		dir->current &= ~(DIR_UP|DIR_LEFT|DIR_RIGHT);
	}
	if (dir->set & DIR_LEFT) {
		object_set_submap(dir->ob, "w-move");
		mapdir_set_motion(dir, -1, 0);
		dir->current |= DIR_LEFT;
		dir->current &= ~(DIR_RIGHT|DIR_UP|DIR_DOWN);
	}
	if (dir->set & DIR_RIGHT) {
		object_set_submap(dir->ob, "e-move");
		mapdir_set_motion(dir, 1, 0);
		dir->current |= DIR_RIGHT;
		dir->current &= ~(DIR_LEFT|DIR_UP|DIR_DOWN);
	}
	dir->set = 0;
}

/* See if the object can move to m:x,y. */
static int
mapdir_can_move(struct mapdir *dir, struct map *dstmap, int x, int y)
{
	struct object *ob = dir->ob;
	struct map *submap = ob->pos->submap;
	struct node *node = &dstmap->map[y][x];
	struct noderef *r;

	if (x < 0 || y < 0 ||
	    x+submap->mapw >= dstmap->mapw ||
	    y+submap->maph >= dstmap->maph) {
		error_set("map boundary");
		return (0);
	}

	if (dir->flags & DIR_PASSTHROUGH) {
		return (1);
	}
	TAILQ_FOREACH(r, &node->nrefs, nrefs) {
		/* Give block precedence over walk. */
		if (ob->pos->z == r->layer &&
		   ((r->flags & NODEREF_WALK)) == 0) {
			error_set("nref block");
			return (0);
		}
	}
	return (1);
}

static void
mapdir_update_moving(struct mapdir *dir, int xmotion, int ymotion)
{
	struct object *ob = dir->ob;
	struct object_position *pos = ob->pos;
	struct map *map = dir->map;
	int moved = 0;
	
	debug(DEBUG_MOVE, "%s: vel %d,%d\n", ob->name, xmotion, ymotion);

	if (!mapdir_can_move(dir, map, pos->x+xmotion, pos->y+ymotion)) {
		debug(DEBUG_BLOCKS, "%s: blocked (%s)\n", ob->name,
		    error_get());
		mapdir_set_motion(dir, 0, 0);
		mapdir_unset(dir, DIR_ALL);
		return;
	}

	if (ymotion < 0) {						/* Up */
		if ((dir->flags & DIR_SCROLLVIEW) && (view->rootmap != NULL))
		    	rootmap_scroll(map, DIR_UP, dir->speed);
		if (ymotion <= -TILEW+dir->speed) {
			dir->moved |= DIR_UP;
			moved |= DIR_UP;
			if (--pos->y < 0)
				pos->y = 0;
		} else {
			if (dir->flags & DIR_SOFTSCROLL)
				mapdir_add_motion(dir, 0, -dir->speed);
			else
				mapdir_set_motion(dir, 0, -TILEH);
		}
	} else if (ymotion > 0) {				     /* Down */
		if ((dir->flags & DIR_SCROLLVIEW) && (view->rootmap != NULL))
		    	rootmap_scroll(map, DIR_DOWN, dir->speed);
		if (ymotion >= TILEW-dir->speed) {
			mapdir_set_motion(dir, 0, 1);
			dir->moved |= DIR_DOWN;
			moved |= DIR_DOWN;
			if (++pos->y >= map->maph)
				pos->y = map->maph-1;
		} else {
			if (dir->flags & DIR_SOFTSCROLL)
				mapdir_add_motion(dir, 0, dir->speed);
			else
				mapdir_set_motion(dir, 0, TILEH);
		}
	}

	if (xmotion < 0) {					    /* Left */
		if ((dir->flags & DIR_SCROLLVIEW) && (view->rootmap != NULL))
		    	rootmap_scroll(map, DIR_LEFT, dir->speed);
		if (xmotion <= -TILEW+dir->speed) {
			mapdir_set_motion(dir, -1, 0);
			dir->moved |= DIR_LEFT;
			moved |= DIR_LEFT;
			if (--pos->x < 0)
				pos->x = 0;
		} else {
			if (dir->flags & DIR_SOFTSCROLL)
				mapdir_add_motion(dir, -dir->speed, 0);
			else
				mapdir_set_motion(dir, -TILEW, 0);
		}
	} else if (xmotion > 0) {				   /* Right */
		if ((dir->flags & DIR_SCROLLVIEW) && (view->rootmap != NULL))
		    	rootmap_scroll(map, DIR_RIGHT, dir->speed);
		if (xmotion >= TILEW-dir->speed) {
			mapdir_set_motion(dir, 1, 0);
			dir->moved |= DIR_RIGHT;
			moved |= DIR_RIGHT;
			if (++pos->x >= map->mapw)
				pos->x = map->mapw-1;
		} else {
			if (dir->flags & DIR_SOFTSCROLL)
				mapdir_add_motion(dir, dir->speed, 0);
			else
				mapdir_set_motion(dir, TILEW, 0);
		}
	}
}

/* Update a map direction and move its associated object if necessary. */
int
mapdir_move(struct mapdir *dir)
{
	struct object_position *pos;
	int x, y;

	pthread_mutex_lock(&dir->ob->lock);
	pos = dir->ob->pos;

	pthread_mutex_lock(&pos->map->lock);
	pthread_mutex_lock(&pos->submap->lock);

	dprintf("%s at %s:[%d,%d]+%dx%d\n", dir->ob->name,
	    OBJECT(pos->submap)->name, pos->x, pos->y,
	    pos->submap->mapw, pos->submap->maph);

	if (pos->x < 0 || pos->y < 0 ||
	    pos->x+pos->submap->mapw >= pos->map->mapw ||
	    pos->y+pos->submap->maph >= pos->map->maph ||
	    pos->z >= pos->map->nlayers) {
		error_set("bad coords");
		goto fail;
	}

	for (y = pos->y; y < pos->y+pos->submap->maph; y++) {
		for (x = pos->x; x < pos->x+pos->submap->mapw; x++) {
			struct node *node = &pos->map->map[y][x];
			struct noderef *r;
			
			TAILQ_FOREACH(r, &node->nrefs, nrefs) {
				if (r->layer != pos->z)
					continue;

				if (r->r_gfx.ymotion == 0 &&
				    r->r_gfx.xmotion == 0) {
					mapdir_update_idle(dir);
				} else {
					mapdir_update_moving(dir,
					    r->r_gfx.xmotion,
					    r->r_gfx.ymotion);
				}
				/*
				 * Other nrefs in this rectangle/layer are
				 * assumed to have the same [xy]motion.
				 */
				goto out;
			}
		}
	}
out:
	pthread_mutex_unlock(&pos->submap->lock);
	pthread_mutex_unlock(&pos->map->lock);
	pthread_mutex_unlock(&dir->ob->lock);
	return (0);
fail:
	pthread_mutex_unlock(&pos->submap->lock);
	pthread_mutex_unlock(&pos->map->lock);
	pthread_mutex_unlock(&dir->ob->lock);
	return (-1);
}

/*
 * Called after a movement, to ensure continuation if necessary, or
 * stop moving.
 */
void
mapdir_postmove(struct mapdir *dir, int *mapx, int *mapy, int moved)
{
	if (dir->clear & DIR_UP) {
		dir->current &= ~DIR_UP;
		object_set_submap(dir->ob, "n-idle");
	}
	if (dir->clear & DIR_DOWN) {
		dir->current &= ~DIR_DOWN;
		object_set_submap(dir->ob, "s-idle");
	}
	if (dir->clear & DIR_LEFT) {
		dir->current &= ~DIR_LEFT;
		object_set_submap(dir->ob, "w-idle");
	}
	if (dir->clear & DIR_RIGHT) {
		dir->current &= ~DIR_RIGHT;
		object_set_submap(dir->ob, "e-idle");
	}
	dir->clear = 0;
	mapdir_set_motion(dir, 0, 0);

	if (dir->current != 0) {
		if (dir->set == 0) {
			/* Continue moving in the same direction. */
			dir->set = moved;
		} else {
			/* Change directions. */
		}

		/* Apply direction changes. */
		mapdir_update_idle(dir);

		dir->current |= dir->set;
		dir->set = 0;
	}
}

